### PR TITLE
feat(platform): getRepoForceRebase -> getBranchForceRebase

### LINKED ISSUE #27050: Convert getRepoForceRebase to a per-branch setting
### Describe the proposed change(s).

The idea behind this feature originally was that rebaseWhen=auto by default, and Renovate keeps branches up to date with their base branch if the repo enforces "branches must be up to date".

Today it's fixed false for Azure, Bitbucket, Code Commit, Gerrit, Gitea.

For Bitbucket Server, it returns true if the repo's default merge strategy is "ff-only".

For GitHub, it returns true if branchProtection.required_status_checks.strict is true. Branch protection is only available on paying/premium accounts.

One problem here is that GitHub's branch protection is per-branch but Renovate only checks it for the default branch. A second problem is that it tries to detect it every run, even if it's a 404 (i.e. 1 API call per repo per run).

If we were to drop this feature altogether then it would mean:
- We keep any automerging branches up-to-date
- Other branches fall back to rebaseWhen=conflicted
- Anyone using GitHub branch protection enforcing up to date would need to configure rebaseWhen=behind-base-branch in Renovate config.

An alternative proposed here would be to change this API to be per-branch (called once per base branch), and only query for it if we have a PR branch with rebaseWhen=auto and not already keeping it up to date due to automerging. We also could cache the result to avoid querying per-run (e.g. cache it for a week if it's true, a day if it's false).

In that case:
- remove getRepoForceRebase() from platform API
- add optional getBranchSettings() to platform API with branchName as only param and `rebaseWhen` as the only return field
- For Bitbucket Server, base it on the repo merge settings
- For GitHub, check branch protection for the specific branch
  * keep and consult a per-brach cache per run in case it's already been answered
  * if getting a 404 then cache that for a day
  * repo cache any 200 response
- For other platforms, omit the function altogether
- Call this function in the code where `platform.getRepoForceRebase()` is called today

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Removes compulsory getRepoForceRebase() platform function and adds optional getBranchForceRebase() function for more granular control.

## Context

Closes #27050

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/bitbucket-server/index.ts [hunk 1]
```diff
@@ -228,9 +228,7 @@ export async function initRepo({
   }
 }
 
-export async function getRepoForceRebase(): Promise<boolean> {
-  logger.debug(`getRepoForceRebase()`);
-
+export async function getBranchForceRebase(): Promise<boolean> {
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.ts:231]:
```suggestion
export async function getBranchForceRebase(_branchName: string): Promise<boolean> {
```
add parameter for later usage?

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 2]
```diff
@@ -534,7 +534,7 @@ describe('modules/platform/bitbucket-server/index', () => {
             .reply(200, {
               mergeConfig: null,
             });
-          const actual = await bitbucket.getRepoForceRebase();
+          const actual = await bitbucket.getBranchForceRebase?.('main');
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:537]:
```suggestion
          const actual = await bitbucket.getBranchForceRebase('main');
```
isn't it there? or use this if it's optional on type, because we know it must exist here.

```suggestion
          const actual = await bitbucket.getBranchForceRebase!('main');
```

File: lib/modules/platform/github/index.ts [hunk 3]
```diff
@@ -309,6 +309,7 @@ async function getBranchProtection(
   }
   const res = await githubApi.getJson<BranchProtection>(
     `repos/${config.repository}/branches/${escapeHash(branchName)}/protection`,
+    { repoCache: true },
```
Comment (viceice) [inline lib/modules/platform/github/index.ts:312]: unrelated change?

**Comment by viceice** [comment]
missing coverage

File: lib/modules/platform/github/index.ts (hunk 3 above)
Comment (rarkins) [reply lib/modules/platform/github/index.ts:312]: Semi-related! Can separate it though if you think that's better

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 4]
```diff
@@ -550,7 +550,7 @@ describe('modules/platform/bitbucket-server/index', () => {
                 defaultStrategy: null,
               },
             });
-          const actual = await bitbucket.getRepoForceRebase();
+          const actual = await bitbucket.getBranchForceRebase?.('main');
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:553]:
```suggestion
          const actual = await bitbucket.getBranchForceRebase!('main');
```

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 5]
```diff
@@ -570,7 +570,7 @@ describe('modules/platform/bitbucket-server/index', () => {
                   },
                 },
               });
-            const actual = await bitbucket.getRepoForceRebase();
+            const actual = await bitbucket.getBranchForceRebase?.('main');
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:573]:
```suggestion
            const actual = await bitbucket.getBranchForceRebase('main');
```

File: lib/modules/platform/gitlab/index.spec.ts [hunk 6]
```diff
@@ -500,7 +500,7 @@ describe('modules/platform/gitlab/index', () => {
           merge_method: 'merge',
         },
       );
-      expect(await gitlab.getRepoForceRebase()).toBeFalse();
+      expect(await gitlab.getBranchForceRebase?.('master')).toBeFalse();
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.spec.ts:503]:
```suggestion
      expect(await gitlab.getBranchForceRebase!('master')).toBeFalse();
```

File: lib/modules/platform/gitlab/index.spec.ts [hunk 7]
```diff
@@ -514,7 +514,7 @@ describe('modules/platform/gitlab/index', () => {
           merge_method: 'ff',
         },
       );
-      expect(await gitlab.getRepoForceRebase()).toBeTrue();
+      expect(await gitlab.getBranchForceRebase?.('master')).toBeTrue();
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.spec.ts:517]:
```suggestion
      expect(await gitlab.getBranchForceRebase!('master')).toBeTrue();
```

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/actions/runs/7774807668/job/21200062565#step:7:1

File: lib/modules/platform/bitbucket-server/index.spec.ts [hunk 8]
```diff
@@ -591,7 +591,7 @@ describe('modules/platform/bitbucket-server/index', () => {
...
```
Comment (viceice) [inline lib/modules/platform/bitbucket-server/index.spec.ts:594]:
```suggestion
            const actual = await bitbucket.getBranchForceRebase!('main');
```
lost one

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
I get this locally:

```
=============================== Coverage summary ===============================
Statements   : 100% ( 39559/39559 )
Branches     : 99.99% ( 12468/12469 )
Functions    : 100% ( 5007/5007 )
Lines        : 100% ( 38988/38988 )
================================================================================
```

**Comment by viceice** [comment]
https://app.codecov.io/gh/renovatebot/renovate/pull/27052/blob/lib/modules/platform/github/index.ts

this file is too big

**Comment by viceice** [comment]
https://app.codecov.io/gh/renovatebot/renovate/pull/27052/blob/lib/modules/platform/github/index.ts#L111

looks like that is the issue, can be simply refactored to use `?.` operator.

**Review by viceice — APPROVED** [review]
